src/main/java/net/minecraftforge/common/smelting/AbstractSmeltingRecipe.java

```diff
@@ -0,0 +1,41 @@
+package net.minecraftforge.common.smelting;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import net.minecraft.item.ItemStack;
+
+/**
+ * Base class for {@code SmeltingRecipe} implementations.
+ */
+public abstract class AbstractSmeltingRecipe implements SmeltingRecipe
+{
+
+    protected final int duration;
+
+    protected AbstractSmeltingRecipe(int duration)
+    {
+        checkArgument(duration > 0);
+        this.duration = duration;
+    }
+
+    @Override
+    public final ItemStack getOutput(ItemStack input)
+    {
+        return matches(input) ? ItemStack.copyItemStack(getOutput0(input)) : null;
+    }
+
+    /**
+     * Get the raw output for this recipe based on the given input. The resulting stack should not be
+     * a defensive copy. This method is only called, if {@code matches} returns true.
+     *
+     * @param input the input stack
+     * @return the raw output stack
+     */
+    protected abstract ItemStack getOutput0(ItemStack input);
```

Comment:
This method / mechanic is mostly a relic from my original implementation of the vanilla recipes where I did not want to implement the whole "search through the list" thing twice but also did not want to execute it twice when looking up recipes (first check `matches` then check `getOutput`, both of which would need to traverse the whole list).
But since Lex suggested splitting the vanilla recipes up anyways I will adjust all this as soon as I have my PC back.